File: src/engine/ListDefinition.ts
import { InkListItem, SerializedInkListItem } from "./InkList";
import { TryGetResult } from "./TryGetResult";

export class ListDefinition {
  public _name: string;
  public _items: Map<SerializedInkListItem, number> | null;
  public _itemNameToValues: Map<string, number>;

  constructor(name: string, items: Map<string, number> | null) {
    this._name = name || "";
    this._items = null;
    this._itemNameToValues = items || new Map();
  }
  get name() {
    return this._name;
  }
  get items() {
    if (this._items == null) {
      this._items = new Map();
      for (let [key, value] of this._itemNameToValues) {
        let item = new InkListItem(this.name, key);
        this._items.set(item.serialized(), value);
      }
    }

    return this._items;
  }

  public ValueForItem(item: InkListItem) {
    if (!item.itemName) return 0;

    let intVal = this._itemNameToValues.get(item.itemName);
    if (typeof intVal !== "undefined") return intVal;
    else return 0;
  }
  public ContainsItem(item: InkListItem) {
    if (!item.itemName) return false;
    if (item.originName != this.name) return false;

    return this._itemNameToValues.has(item.itemName);
  }
  public ContainsItemWithName(itemName: string) {
    return this._itemNameToValues.has(itemName);
  }
  public TryGetItemWithValue(
    val: number,
    /* out */ item: InkListItem
  ): TryGetResult<InkListItem> {
    for (let [key, value] of this._itemNameToValues) {
      if (value == val) {
        item = new InkListItem(this.name, key);
        return { result: item, exists: true };
      }
    }

    item = InkListItem.Null;
    return { result: item, exists: false };
  }

  public TryGetValueForItem(
    item: InkListItem,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    /* out */ intVal: number
  ): TryGetResult<number> {
    if (!item.itemName) return { result: 0, exists: false };
    let value = this._itemNameToValues.get(item.itemName);

    if (!value) return { result: 0, exists: false };
    return { result: value, exists: true };
  }
}

File: ink_dart/lib/src/engine/list_definition.dart
import 'ink_list.dart';

/// Definition of a single LIST in Ink.
///
/// Contains the mapping of item names to their integer values.
class ListDefinition {
  /// The name of the list (e.g., "Colors", "Items").
  final String name;

  /// Raw items stored as serialized keys to values.
  final Map<String, int> _items;

  ListDefinition(this.name, Map<String, int> items) : _items = items;

  /// All items in this definition.
  Map<String, int> get items => _items;

  /// Get an item by value.
  ({bool exists, InkListItem? result}) tryGetItemWithValue(
    int value,
    InkListItem defaultValue,
  ) {
    for (final entry in _items.entries) {
      if (entry.value == value) {
        return (exists: true, result: InkListItem(name, entry.key));
      }
    }
    return (exists: false, result: defaultValue);
  }

  /// Get value for an item.
  ({bool exists, int? result}) tryGetValueForItem(InkListItem item) {
    final key = item.itemName;
    if (key != null && _items.containsKey(key)) {
      return (exists: true, result: _items[key]);
    }
    return (exists: false, result: null);
  }

  /// Check if this definition contains an item.
  bool containsItem(InkListItem item) {
    final key = item.itemName;
    return key != null && _items.containsKey(key);
  }

  /// Check if this definition contains an item by name.
  bool containsItemWithName(String itemName) {
    return _items.containsKey(itemName);
  }

  @override
  String toString() => 'ListDefinition($name, ${_items.length} items)';
}
